packages/vaadin-overlay/src/vaadin-overlay-position-mixin.js

```diff
@@ -80,25 +81,58 @@ export const PositionMixin = (superClass) =>
     static get observers() {
       return [
         '__positionSettingsChanged(positionTarget, horizontalAlign, verticalAlign, noHorizontalOverlap, noVerticalOverlap)',
-        '__overlayOpenedChanged(opened)',
+        '__overlayOpenedChanged(opened, positionTarget)',
       ];
     }
 
     constructor() {
       super();
 
-      this.__boundUpdatePosition = this._updatePosition.bind(this);
+      this._updatePosition = this._updatePosition.bind(this);
     }
 
-    __overlayOpenedChanged(opened) {
-      // Toggle the event listeners that cause the overlay to update its position
-      ['scroll', 'resize'].forEach((eventName) => {
-        if (opened) {
-          window.addEventListener(eventName, this.__boundUpdatePosition);
-        } else {
-          window.removeEventListener(eventName, this.__boundUpdatePosition);
-        }
+    connectedCallback() {
+      super.connectedCallback();
+
+      if (this.opened) {
+        this.__addUpdatePositionEventListeners();
+      }
+    }
+
+    disconnectedCallback() {
+      super.disconnectedCallback();
+      this.__removeUpdatePositionEventListeners();
+    }
+
+    /** @private */
+    __addUpdatePositionEventListeners() {
+      window.addEventListener('resize', this._updatePosition);
+
+      this.__positionTargetAncestorRootNodes = getAncestorRootNodes(this.positionTarget);
+      this.__positionTargetAncestorRootNodes.forEach((node) => {
+        node.addEventListener('scroll', this._updatePosition, true);
       });
+    }
+
+    /** @private */
+    __removeUpdatePositionEventListeners() {
+      window.removeEventListener('resize', this._updatePosition);
+
+      if (this.__positionTargetAncestorRootNodes) {
+        this.__positionTargetAncestorRootNodes.forEach((node) => {
+          node.removeEventListener('scroll', this._updatePosition, true);
+        });
+        this.__positionTargetAncestorRootNodes = null;
+      }
+    }
+
+    __overlayOpenedChanged(opened, positionTarget) {
+      // Toggle the event listeners that cause the overlay to update its position
+      if (opened && positionTarget) {
+        this.__addUpdatePositionEventListeners();
+      } else {
+        this.__removeUpdatePositionEventListeners();
```

Comment:
nit: Should we always run `__removeUpdatePositionEventListeners` as the first thing to cover the (very unlikely) case that the `positionTarget` changes from one element to another one that generates a different set of ancestor root nodes? In that case, the listeners may not get cleaned up from the old ancestor nodes.

Reply:
Good point, this case slipped my mind. I'm sure it makes sense to address that in this PR because potential memory leaks can be difficult to reveal later.

Reply:
Fixed. https://github.com/vaadin/web-components/pull/3932/commits/acfc6ff6517eaf673e702670901bb88e59a48430